substreams/ethereum-nabla/src/modules/map_components.rs

```diff
@@ -0,0 +1,195 @@
+        contracts: vec![],
+        static_att: vec![],
+        protocol_type: Option::from(ProtocolType {
+            name: ComponentType::Router.into(),
+            financial_type: FinancialType::Swap.into(),
+            attribute_schema: vec![],
+            implementation_type: ImplementationType::Custom.into(),
+        }),
+    })
+}
+
+fn initialize_swap_pool(
+    tx: &TransactionTrace,
+    pool_address: Vec<u8>,
+) -> Result<ProtocolComponent, String> {
+    let pool_id = pool_address.to_hex();
+
+    let asset = abi::swap_pool::functions::Asset {}
+        .call(pool_address.clone())
+        .ok_or_else(|| format!("Failed to retrieve 'asset' for pool {:?}", pool_id))?;
+
+    let name = abi::swap_pool::functions::Name {}
+        .call(pool_address.clone())
+        .ok_or_else(|| format!("Failed to retrieve 'name' for pool {:?}", pool_id))?;
+
+    let symbol = abi::swap_pool::functions::Symbol {}
+        .call(pool_address.clone())
+        .ok_or_else(|| format!("Failed to retrieve 'symbol' for pool {:?}", pool_id))?;
+
+    let attributes = vec![
+        Attribute {
+            name: "name".to_string(),
+            value: name.into(),
+            change: ChangeType::Creation.into(),
+        },
+        Attribute {
+            name: "symbol".to_string(),
+            value: symbol.into(),
+            change: ChangeType::Creation.into(),
+        },
+    ];
+
+    Ok(ProtocolComponent {
+        id: pool_id,
+        tokens: vec![asset],
+        tx: Some(tx.into()),
+        change: i32::from(ChangeType::Creation),
+        contracts: vec![],
+        static_att: attributes,
+        protocol_type: Option::from(ProtocolType {
+            name: ComponentType::SwapPool.into(),
+            financial_type: FinancialType::Swap.into(),
+            attribute_schema: vec![],
+            implementation_type: ImplementationType::Custom.into(),
+        }),
+    })
+}
+
+fn detect_asset_registered_event(log: &Log) -> Option<abi::nabla_portal::events::AssetRegistered> {
+    (log.address == NABLA_PORTAL.contract_address)
+        .then(|| abi::nabla_portal::events::AssetRegistered::match_and_decode(log))?
+}
+
+fn to_component(
+    tx: &TransactionTrace,
+    asset_registered: abi::nabla_portal::events::AssetRegistered,
+) -> Result<[ProtocolComponent; 2], String> {
+    let router_id = asset_registered.router.to_hex();
+    let asset_id = asset_registered.asset.to_hex();
+
+    let pool_address = abi::nabla_router::functions::PoolByAsset { param0: asset_registered.asset }
+        .call(asset_registered.router.clone())
+        .ok_or_else(|| {
+            format!(
+                "Failed to retrieve pool_address for asset {} from the router {}",
+                asset_id, router_id
+            )
+        })?;
+
+    let router = initialize_router(tx, asset_registered.router)?;
+    let swap_pool = initialize_swap_pool(tx, pool_address.clone())?;
+    Ok([router, swap_pool])
+}
```

Comment:
all `initialize_` functions only initialize immutable attributes, not mutable state, that is dealt with separately later on. This allows adhering to the single responsibility principle, and allows returning only `BlockTransactionProtocolComponents`, instead of `BlockChanges` (which include mutable attribute data) at this stage.